Terminal.Gui/ConsoleDrivers/WindowsDriver/WindowsDriver.cs

```diff
@@ -1177,6 +198,73 @@ public override void SendKeys (char keyChar, ConsoleKey key, bool shift, bool al
         }
     }
 
+    private readonly ManualResetEventSlim _waitAnsiResponse = new (false);
+    private readonly CancellationTokenSource _ansiResponseTokenSource = new ();
+
+    /// <inheritdoc/>
+    public override string WriteAnsiRequest (AnsiEscapeSequenceRequest ansiRequest)
+    {
+        if (_mainLoopDriver is null)
+        {
+            return string.Empty;
+        }
+
+        try
+        {
+            lock (ansiRequest._responseLock)
+            {
+                ansiRequest.ResponseFromInput += (s, e) =>
+                                                 {
+                                                     Debug.Assert (s == ansiRequest);
+                                                     Debug.Assert (e == ansiRequest.Response);
+
+                                                     _waitAnsiResponse.Set ();
+                                                 };
+
```

Comment:
This code is in all 3 drivers. It looks like you block 100ms to let the sent request get processed and then 'force run' the input stream for a bit, then look for a response? 

Is it possible for you to explain in some pseudocode how it is working?

This PR is enormous and has so much in it that isn't related to ansi response send/capture it's hard to understand what is going on.

I've tried to standardise input processing with AnsiParser<T> to replace NetEvents and similar code in the other drivers. But now all the drivers have huge changes in input handling its going to be difficult to retro fit.

I don't think we need 2 solutions to ansi send. I'd be happy if we just did 

- get rid of all the locking instead use BlockingCollections
- standardise the parsing of request across all 3 drivers into a single class (i.e. AnsiRequestParser). 

We can still do the blocking on response but the class that identifies responses in input streams needs to be centralised into a single class

Reply:
> This code is in all 3 drivers. It looks like you block 100ms to let the sent request get processed and then 'force run' the input stream for a bit, then look for a response?
> 
> Is it possible for you to explain in some pseudocode how it is working?
> 

It's a ManualResetEventSlim like it's used in the MainLoop that wait to be signaled. It only wait until has a response or until two retries and return immediately.

> This PR is enormous and has so much in it that isn't related to ansi response send/capture it's hard to understand what is going on.
> 
I really refactored the `CursesDriver` to leverage using non blocking input, like I did in the `WindowsDriver`.

> I've tried to standardise input processing with AnsiParser to replace NetEvents and similar code in the other drivers. But now all the drivers have huge changes in input handling its going to be difficult to retro fit.
> 
I think that isn't very difficult to implement, beside the changes.

> I don't think we need 2 solutions to ansi send. I'd be happy if we just did
> 
> * get rid of all the locking instead use BlockingCollections

I'm open on that but I need some explanation of that. I tough using the `ManualResetEventSlim` was a good idea, but I would like to know about alternatives, of course.

> * standardise the parsing of request across all 3 drivers into a single class (i.e. AnsiRequestParser).
> 
The `NetDriver` and the `CursesDriver` relies in the `EscSeqUtil.DecodeEscSeq` method because they are handling key and mouse events, but the `WindowsDriver` relies on the `WindowsConsole.ReadConsoleInput` because any other key that it isn't a normal key or a mouse event, is handled in this method.

> We can still do the blocking on response but the class that identifies responses in input streams needs to be centralised into a single class

Well due to my above explanation `NetDriver` and the `CursesDriver` relies in the `EscSeqUtil.DecodeEscSeq` and the `WindowsDriver` relies on the `WindowsConsole.ReadConsoleInput`. So, I don't know if they can be in a single class. But I understand your concern.